x/expiration/keeper/keeper.go

```diff
@@ -0,0 +1,565 @@
+	}
+	if oldExpiration == nil {
+		return types.ErrExtendExpiration.Wrapf("%s: %s", types.ErrNotFound.Error(), expiration.ModuleAssetId)
+	}
+
+	// Validate expiration time is after old expiration time
+	if oldExpiration.Time.After(expiration.Time) {
+		return types.ErrExtendExpiration.Wrapf("new expiration time [%s] must be after old expiration time [%s]",
+			expiration.Time.UTC(), oldExpiration.Time.UTC())
+	}
+	// Validate owners are the same
+	if expiration.Owner != oldExpiration.Owner {
+		return types.ErrNewOwnerNoMatch.Wrapf("new owner [%s] and old owner [%s] do not match",
+			expiration.Owner, oldExpiration.Owner)
+	}
+
+	// Marshal expiration record and store
+	store := ctx.KVStore(k.storeKey)
+	b, err := k.cdc.Marshal(&expiration)
+	if err != nil {
+		return types.ErrExtendExpiration.Wrapf("%s: %v", types.ErrUnmarshal.Error(), err)
+	}
+	store.Set(key, b)
+
+	// emit Extend event
+	k.emitEvent(ctx, types.NewEventExpirationExtend(expiration.ModuleAssetId))
+
+	return nil
+}
+
+// InvokeExpiration invokes an expiration message through the MsgServiceRouter.
+// The expiration message is removed when the message invocation is successful.
+//
+// Caution: Do NOT call InvokeExpiration from within a module to delete a self-owned asset.
+// This will lead to an infinite loop as expiration messages in this method are routed through
+// the message router. Use RemoveExpiration instead.
+func (k Keeper) InvokeExpiration(ctx sdk.Context, moduleAssetID string, refundTo sdk.AccAddress) error {
+	// lookup expiration
+	expiration, err := k.GetExpiration(ctx, moduleAssetID)
+	if err != nil {
+		return types.ErrInvokeExpiration.Wrap(err.Error())
+	}
+	if expiration == nil {
+		return types.ErrInvokeExpiration.Wrapf("%s [%s]", types.ErrNotFound.Error(), expiration.ModuleAssetId)
+	}
+
+	// unpack expiration message
+	var msg sdk.Msg
+	if err = k.cdc.UnpackAny(&expiration.Message, &msg); err != nil {
+		return types.ErrInvokeExpiration.Wrapf("failed to unpack msg: %v", err)
+	}
+
+	// route message to module
+	handler := k.router.Handler(msg)
+	if handler == nil {
+		return types.ErrInvokeExpiration.Wrap(
+			types.ErrMsgHandler.Wrapf("no message handler found for %q", sdk.MsgTypeURL(msg)).Error())
+	}
+	r, err := handler(ctx, msg)
+	if err != nil {
+		return types.ErrInvokeExpiration.Wrap(
+			types.ErrMsgHandler.Wrapf("message %s: %v", sdk.MsgTypeURL(msg), err).Error())
+	}
+	// Handler should always return non-nil sdk.Result.
+	if r == nil {
+		return types.ErrInvokeExpiration.Wrap(
+			types.ErrMsgHandler.Wrapf("got nil sdk.Result for message %q", msg).Error())
+	}
+
+	// refund deposit from expiration module account to depositor
+	if err := k.refundDeposit(ctx, expiration.Deposit, refundTo); err != nil {
+		return types.ErrInvokeExpiration.Wrapf("failed to refund deposit to [%s]: %v", refundTo, err)
+	}
+
+	// remove record from store
+	if err := k.removeExpiration(ctx, moduleAssetID); err != nil {
```

Comment:
This might not be of concern for this PR. I was wondering if there is a case where this method fails (and will always fail...due to some issue) and never reaches the `removeExpiration` function.  The expiration will remain in the keeper and someone could keep calling `InvokeExpiration` on that expiration since it will always show up as a valid one to call when queried.